compiler/rustc_mir/src/transform/early_otherwise_branch.rs

```diff
@@ -26,7 +27,7 @@ pub struct EarlyOtherwiseBranch;
 
 impl<'tcx> MirPass<'tcx> for EarlyOtherwiseBranch {
     fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
-        if tcx.sess.opts.debugging_opts.mir_opt_level < 2 {
+        if tcx.sess.opts.debugging_opts.mir_opt_level < min_mir_opt_level::EARLY_OTHERWISE {
```

Comment:
I don't think I'm a fan of this change (adding the `min_mir_opt_level` constants). The `mir-opt-level` values are organized around the effects they have on code and compiler performance not so much about what specific optimizations are applied (that's why when the inliner is more stable/performant it might be enabled at `mir-opt-level=1`). 

However, I do see why you made this change and I think a simpler thing to do might just be to add an explicit check for `-Zinstrument-coverage` to `Inliner.rs` and if that flag is enabled, skip the inliner at any `mir-opt-level`. 

`-Zmir-opt-level` doesn't commit us to running any specific optimizations so that still seems valid to me.

Reply:
> I don't think I'm a fan of this change

I'm not sure I understand what you're objecting to, specifically. The logic and control is still there. The decisions about what levels apply what optimizations is more clear, and more centrally defined and accessible.

**_BUT_**, I have been reviewing this again, and I do see some issues; some minor nits, but one big problem actually:

If someone decides to promote inlining to mir-opt-level 1, that's the default opt level, and now `-Zinstrument-coverage` will fail without any other flags! I agree that's bad, so in effect, the `error` in `config.rs` will (without other changes) prevent promoting inling to mir-opt-level 1!

I agree that's bad.

We'd have to remove the flag check and error, and do something like you suggest, that is, add the check in the inline pass. 

I was hoping to avoid scattering coverage-specific exceptions in unrelated parts of the codebase. (If the code changes, it's easy to miss little unrelated piece of logic like this.) It would be nicer if each optimization could be turned on or off via session option, rather than having them grouped together under these hardcoded numbers.

There is a way to still get the advantages of centralized and accessible optimizations by level, but also allows optimizations to be turned on or off granularly, during command line processing:

In `rustc_session` add a new field to the `Options` struct (using the `bitflags!` pattern currently applied to `SanitizerSet`):

```rust
    mir_opts: MirOpt,
...

bitflags! {
    #[derive(Default, Encodable, Decodable)]
    pub struct MirOpt: u16 {
        const CONST_PROP_BASIC = 1 << 0;
        const RETURN_PLACE    = 1 << 1;

        const MIR_OPT_LEVEL_1 = (1 << 2) - 1;

        const CONST_PROP_COPIES_AND_MOVES  = 1 << 2;
        const DEST_PROP  = 1 << 3;
        const EARLY_OTHERWISE  = 1 << 4;
        const INLINE  = 1 << 5;
        const MATCH_BRANCH  = 1 << 6;

        const MIR_OPT_LEVEL_2 = (1 << 7) - 1;

        const CONST_IDENTITY  = 1 << 7;
        const RETURN_TERM  = 1 << 8;
        const UNREACHABLE_PROP  = 1 << 9;

        const MIR_OPT_LEVEL_3 = (1 << 10) - 1;
    }
}
```

and then, at the top of `build_session_options()`, enable each option based on the selected mir-opt-level:

```rust
    if debugging_opts.mir_opt_level >= 1 {
        mir_opts.insert(MirOpt::MIR_OPT_LEVEL_1);
    }
    if debugging_opts.mir_opt_level >= 2 {
        mir_opts.insert(MirOpt::MIR_OPT_LEVEL_2);
    }
    if debugging_opts.mir_opt_level >= 3 {
        mir_opts.insert(MirOpt::MIR_OPT_LEVEL_3);
    }
```

Then I can change my command line check to the following, and GRANULARLY disable the inline pass, only, regardless of what `level` it's using. AND I can print a warning, to notify the developer that, if they are expecting inlining by setting `-Zmir-opt-level=2`, it is disabled.

```rust
        if mir_opts.contains(MirOpt::INLINE) {
            mir_opts.remove(MirOpt::INLINE);
            if MirOpt::INLINE & MirOpt::MIR_OPT_LEVEL_1 == 0 {
                // Warn only if inlining is not part of the default optimizations
                early_warn(... "`-Z instrument-coverage` disables MIR optimization pass `Inline`");
...
```

And other similar checks, like that in `inline.rs` change to:

```rust
        if ! tcx.sess.opts.mir_opts.contains(min_mir_opt_level::INLINE) {
            return;
        }
```

As I said, I'm not sure what your initial objections were, and this new proposal still centralizes where optimizations relate to mir-opt-level. I think that's a good thing. Let me know your thoughts on this alternative.

Reply:
Sorry, I should have expanded on my position a bit more. 

I think the reason I don't like this approach is that I think it makes `rustc_session` far too knowledgeable about which MIR optimizations are run and even what parts of those optimizations are run. This seems strange to me because the option `rustc_session` is responsible for configuring is intentionally vague about what optimizations are run exactly. For instance:

- `-Zmir-opt-level=0` => don't run any MIR transformations that aren't required to do codegen
- `-Zmir-opt-level=1` => run any MIR optimizations as long as they don't hurt debugging or compilation time 
- `-Zmir-opt-level=2` => run any MIR optimizations as long as they don't hurt compilation time "too much"
- `-Zmir-opt-level=3` => run any MIR optimizations without regard to debugability or compilation time cost

Since the flags aren't specified as running any particular optimizations, it seems weird to me that `rustc_session` needs to know what optimizations are available to be run. Especially since some optimizations are tuned to be more or less aggressive depending on the optimization level requested.

I guess another way to look at it is that the details of MIR optimizations are currently encapsulated in `rustc_mir::transform` and I think that's a good thing. I would prefer that we don't expose what optimizations are available or configured to run outside of `rustc_mir`. 

> I was hoping to avoid scattering coverage-specific exceptions in unrelated parts of the codebase. (If the code changes, it's easy to miss little unrelated piece of logic like this.)

I think this is generally a good goal but a lot of the MIR optimizations are already fairly context sensitive as to what the build flags are and what kind of MIR `Body` they are operating on. Adding the check for `-Zinstrument-coverage` to the Inliner doesn't seem like a big deal in this case. 

Of course, the option would be to fix the Inliner so that we can inline things even when code coverage is enabled. If you think there's a way to make that work, I would be willing to work on implementing that. 🙂

Reply:
Done.

I get that my proposal is a bigger change than just fixing the ICE (and the debate is more philosophical than practical at this point), so I defer to your judgement here.

> Of course, the option would be to fix the Inliner so that we can inline things even when code coverage is enabled. If you think there's a way to make that work, I would be willing to work on implementing that

I doubt it's worth doing. I don't see a benefit in optimizing code while also injecting coverage counters. There might be exceptions, but generally it seems like solving all of the problems this can introduce, while also making sure the coverage results are still accurate, is a lot of hoops to jump through, and unnecessary if we accept the limitation that coverage is a feature meant to be used in non-production environments.

That's just my opinion though.

Reply:
That makes complete sense to me. Thanks for your feedback!